cryptarchia/cryptarchia.py

```diff
@@ -250,57 +290,60 @@ def on_block(self, block: BlockHeader):
-        else:
-            # we have re-org'd, therefore we must roll back out ledger state and
-            # re-apply blocks from the new chain
-            ledger_state = self.genesis_state.copy()
-            for block in new_chain.blocks:
-                ledger_state.apply(block)
-
-            self.ledger_state = ledger_state
-            self.local_chain = new_chain
+        new_state = self.ledger_state[block.parent].copy()
+        new_state.apply(block)
+        self.ledger_state[block.id()] = new_state
 
     # Evaluate the fork choice rule and return the block header of the block that should be the head of the chain
     def fork_choice(self) -> Chain:
         return maxvalid_bg(
             self.local_chain, self.forks, k=self.config.k, s=self.config.s
         )
 
+    def tip(self) -> BlockHeader:
+        return self.local_chain.tip()
+
     def tip_id(self) -> Id:
         if self.local_chain.length() > 0:
             return self.local_chain.tip().id()
         else:
-            return self.ledger_state.block
+            return self.genesis_state.block
 
+    def get_last_valid_state(self, chain: Chain, slot: Slot) -> LedgerState:
+        for block in reversed(chain.blocks):
+            if block.slot < slot:
+                return self.ledger_state[block.id()]
 
-@dataclass
-class EpochState:
-    # for details of snapshot schedule please see:
-    # https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#epoch-structure
-
-    # The stake distribution snapshot is taken at the beginning of the previous epoch
-    stake_distribution_snapshot: LedgerState
+        return self.genesis_state
 
-    # The nonce snapshot is taken 7k/f slots into the previous epoch
-    nonce_snapshot: LedgerState
-
-    def verify_commitment_is_old_enough_to_lead(self, commitment: Id) -> bool:
-        return self.stake_distribution_snapshot.verify_committed(commitment)
-
-    def total_stake(self) -> int:
-        """Returns the total stake that will be used to reletivize leadership proofs during this epoch"""
-        return self.stake_distribution_snapshot.total_stake
+    def compute_epoch_state(self, epoch: Epoch, chain: Chain) -> EpochState:
+        # stake distribution snapshot happens at the beginning of the previous epoch,
+        # i.e. for epoch e, the snapshot is taken at the last block of epoch e-2
+        stake_snapshot_slot = Slot((epoch.epoch - 1) * self.config.epoch_length())
+        stake_distribution_snapshot = self.get_last_valid_state(
+            chain, stake_snapshot_slot
+        )
+        # nonce snapshot is taken 7k/f slots into the previous epoch
+        # i.e. for epoch e, the snapshot is taken at the block at slot floor(7k/f) of epoch e-1
+        nonce_slot = Slot(
+            int(floor(7 * self.config.k / self.config.active_slot_coeff))
+            + stake_snapshot_slot.absolute_slot
+        )
```

Comment:
yeah I think putting the epoch periods in the config would make this code clearer